blob/blob.go

```diff
@@ -16,42 +19,70 @@ import (
 	"github.com/celestiaorg/celestia-node/share"
 )
 
+//nolint:unused
 var errEmptyShares = errors.New("empty shares")
 
-// The Proof is a set of nmt proofs that can be verified only through
-// the included method (due to limitation of the nmt https://github.com/celestiaorg/nmt/issues/218).
-// Proof proves the WHOLE namespaced data to the row roots.
-// TODO (@vgonkivs): rework `Proof` in order to prove a particular blob.
-// https://github.com/celestiaorg/celestia-node/issues/2303
-type Proof []*nmt.Proof
-
-func (p Proof) Len() int { return len(p) }
+// Proof constructs the proof of a blob to the data root.
+type Proof struct {
+	// ShareToRowRootProof the proofs of the shares to the row roots they belong to.
+	// If the blob spans across multiple rows, then this will contain multiple proofs.
+	ShareToRowRootProof []*tmproto.NMTProof
+	// RowProof the proofs of the row roots containing the blob shares
+	// to the data root.
+	RowProof coretypes.RowProof
+}
 
-// equal is a temporary method that compares two proofs.
-// should be removed in BlobService V1.
-func (p Proof) equal(input Proof) error {
-	if p.Len() != input.Len() {
-		return ErrInvalidProof
+// Verify takes a blob and a data root and verifies if the
+// provided blob was committed to by the data root.
+func (p *Proof) Verify(blob *Blob, dataRoot []byte) (bool, error) {
+	blobCommitment, err := types.CreateCommitment(ToAppBlobs(blob)[0])
+	if err != nil {
+		return false, err
+	}
+	if !blob.Commitment.Equal(blobCommitment) {
```

Comment:
Here we are checking if the blob commitment we generate using the given blob matches the commitment inside the given blob? 

What does this have to do with the proof then (re: error message)? wouldn't it just be a malformed blob?

Reply:
that's a malformed blob, yes. should I create a new error type?

Reply:
Not sure if a whole new error type is necessary unless we have multiple places where a malformed blob can be detected. regardless, we need the err msg to be accurate

Reply:
Yes bc this error doesn't have anything to do with the proof, just the blob

Reply:
[3630f33](https://github.com/celestiaorg/celestia-node/pull/3610/commits/3630f332e858c72ae0445e1abb9871b0fcc1e946)